src/core/qgstessellator.cpp

```diff
@@ -436,21 +459,66 @@ void QgsTessellator::addPolygon( const QgsPolygon &polygon, float extrusionHeigh
 
   if ( pCount == 4 && polygon.numInteriorRings() == 0 )
   {
+    std::vector<p2t::Point *> polyline;
+    if ( mAddTextureCoords )
+    {
+      std::unique_ptr<QMatrix4x4> toNewBase, toOldBase;
+      if ( !mNoZ && pNormal != QVector3D( 0, 0, 1 ) )
+      {
+        // this is not a horizontal plane - need to reproject the polygon to a new base so that
+        // we can do the triangulation in a plane
+
+        QVector3D pXVector, pYVector;
+        _normalVectorToXYVectors( pNormal, pXVector, pYVector );
+
+        // so now we have three orthogonal unit vectors defining new base
+        // let's build transform matrix. We actually need just a 3x3 matrix,
+        // but Qt does not have good support for it, so using 4x4 matrix instead.
+        toNewBase.reset( new QMatrix4x4(
+                           pXVector.x(), pXVector.y(), pXVector.z(), 0,
+                           pYVector.x(), pYVector.y(), pYVector.z(), 0,
+                           pNormal.x(), pNormal.y(), pNormal.z(), 0,
+                           0, 0, 0, 0 ) );
+
+        // our 3x3 matrix is orthogonal, so for inverse we only need to transpose it
+        toOldBase.reset( new QMatrix4x4( toNewBase->transposed() ) );
+      }
+
+      const QgsPoint ptStart( exterior->startPoint() );
+      const QgsPoint pt0( QgsWkbTypes::PointZ, ptStart.x(), ptStart.y(), std::isnan( ptStart.z() ) ? 0 : ptStart.z() );
+
+      const float scaleX = !mBounds.isNull() ? 10000.0 / mBounds.width() : 1.0;
+      const float scaleY = !mBounds.isNull() ? 10000.0 / mBounds.height() : 1.0;
+
+      // subtract ptFirst from geometry for better numerical stability in triangulation
+      // and apply new 3D vector base if the polygon is not horizontal
+
+      std::unique_ptr<QgsPolygon> polygonNew( _transform_polygon_to_new_base( polygon, pt0, toNewBase.get(), scaleX, scaleY ) );
+      QList< std::vector<p2t::Point *> > polylinesToDelete;
+      QHash<p2t::Point *, float> z;
+
+      _ringToPoly2tri( qgsgeometry_cast< const QgsLineString * >( polygonNew->exteriorRing() ), polyline, mNoZ ? nullptr : &z );
```

Comment:
We are working with a single triangle here, so do we need to run triangulation here? Can't we just use the coordinates from `polygonNew` ?